{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Barnesville/schools/010/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Barnesville/schools/010/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Barnesville/schools/010/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Barnesville/schools/040/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Barnesville/schools/040/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Barnesville/schools/040/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"}

