
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Ulen%20Hitterdahl/schools/002/segments/Guardian?sendFlatSpecs=true","destination":"Guardians", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Ulen%20Hitterdahl/schools/002/segments/Schedule?sendFlatSpecs=true","destination":"Schedules", "path":"Data"},
{"source":"http://127.0.0.1:<!port!>/uff/1.0/districts/Ulen%20Hitterdahl/schools/002/segments/Student_Enrolled?sendFlatSpecs=true","destination":"Students", "path":"Data"}


