// serverProfiles/mattExtDotEx/mattExtDotEx.js
'use strict';
var qtools = require('qtools'),
	qtools = new qtools(module),
	events = require('events'),
	util = require('util');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {
	events.EventEmitter.call(this);
	this.forceEvent = forceEvent;
	this.args = args;
	this.metaData = {};
	this.addMeta = function(name, data) {
		this.metaData[name] = data;
	}

	// 	qtools.validateProperties({
	// 		subject: args,
	// 		targetScope: this, //will add listed items to targetScope
	// 		propList: [
	// 			{
	// 				name: 'placeholder',
	// 				optional: true
	// 			}
	// 		]
	// 	});

	var self = this,
		forceEvent = function(eventName, outData) {
			this.emit(eventName, {
				eventName: eventName,
				data: outData
			});
		};


	//LOCAL FUNCTIONS ====================================
	var mapGenerator=require('../lib/uffRemapNames.js');
	var mappingTable = new mapGenerator().mappingTable;


// 	var tableNameMap={
// 		AddressContact:'ext.exContact',
// 		Course:'ext.exCourse',
// 		UserBase:'ext.exUserBase',
// 		Section:'ext.exSection',
// 		SectionStaff:'ext.exSectionStaff',
// 		SectionStudent:'ext.exSectionStudent',
// 		StudentAttendance:'ext.exStudentAttendance',
// 		StudentBase:'ext.exStudentBase',
// 		StudentDiscipline:'ext.exStudentDiscipline',
// 		StudentEnrollment:'ext.exStudentEnrollment',
// 		StudentGpa:'ext.exStudentGPA',
// 		StudentGradeLevels:'ext.exStudentGrades'
// 	}

	var tableNameMap={
		AddressContact:'uff.exAddressContact',
		Course:'uff.exCourse',
		UserBase:'uff.exUserBase',
		Section:'uff.exSection',
		SectionStaff:'uff.exSectionStaff',
		SectionStudent:'uff.exSectionStudent',
		StudentAttendance:'uff.exStudentAttendance',
		StudentBase:'uff.exStudentBase',
		StudentDiscipline:'uff.exStudentDiscipline',
		StudentEnrollment:'uff.exStudentEnrollment',
		StudentGpa:'uff.exStudentGPA',
		StudentGradeLevels:'uff.exStudentGrades',
		
		Address_Contact:'uff.exAddressContact',
	//	Course:'uff.exCourse',
		User_Base:'uff.exUserBase',
	//	Section:'uff.exSection',
		Section_Staff:'uff.exSectionStaff',
		Section_Student:'uff.exSectionStudent',
		Student_Attendance:'uff.exStudentAttendance',
		Student_Base:'uff.exStudentBase',
		Student_Discipline:'uff.exStudentDiscipline',
		Student_Enrollment:'uff.exStudentEnrollment',
		Student_gpa:'uff.exStudentGPA',
		Student_gradeLevels:'uff.exStudentGrades'
	}

// 	var tableNameMap={
// 		AddressContact:'ext.ExAddressContact',
// 		Course:'ext.ExCourse',
// 		UserBase:'ext.ExUserBase',
// 		Section:'ext.ExSection',
// 		SectionStaff:'ext.ExSectionStaff',
// 		SectionStudent:'ext.ExSectionStudent',
// 		StudentAttendance:'ext.ExStudentAttendance',
// 		StudentBase:'ext.ExStudentBase',
// 		StudentDiscipline:'ext.ExStudentDiscipline',
// 		StudentEnrollment:'ext.ExStudentEnrollment',
// 		StudentGpa:'ext.ExStudentGPA',
// 		StudentGradeLevels:'ext.ExStudentGrades',
// 		
// 		Address_Contact:'ext.ExAddressContact',
// 	//	Course:'ext.ExCourse',
// 		User_Base:'ext.ExUserBase',
// 	//	Section:'ext.ExSection',
// 		Section_Staff:'ext.ExSectionStaff',
// 		Section_Student:'ext.ExSectionStudent',
// 		Student_Attendance:'ext.ExStudentAttendance',
// 		Student_Base:'ext.ExStudentBase',
// 		Student_Discipline:'ext.ExStudentDiscipline',
// 		Student_Enrollment:'ext.ExStudentEnrollment',
// 		Student_gpa:'ext.ExStudentGPA',
// 		Student_gradeLevels:'ext.ExStudentGrades'
// 	}

	//METHODS AND PROPERTIES ====================================

	this.mapDistrictAccessName = function(dictionaryName) {
		return "SSI_XDI_" + dictionaryName;
	}

	this.mapSegmentAccessName = function(dictionaryName) {

		var mapResult=tableNameMap[dictionaryName];
		if (typeof(mapResult)!='undefined'){
			return mapResult;
		}
		else{
			return dictionaryName;
		}
	}
	
	this.mapQueryNames=function(query){
		return qtools.templateReplace({template:query, replaceObject:tableNameMap});
	}


	this.mapFieldNames = function(inData, stripNulls) {
		var outObj = {};
		for (var i in inData) {
			var element = inData[i],
				replacement = mappingTable[i];

			if (stripNulls && (element && element.toLowerCase() == 'null') || qtools.toType(element) == 'null') {
				element = '';
			}

			if (typeof (replacement) != 'undefined') {
				outObj[replacement] = element;
			} else {
				outObj[replacement] = element;
			}
		}
		return outObj;
	}

	this.mapFieldNamesList = function(inArray, stripNulls) {
		stripNulls = (typeof (stripNulls) == 'undefined') ? false : stripNulls;
		var outArray = [];
		for (var i = 0, len = inArray.length; i < len; i++) {
			var element = inArray[i];
			outArray.push(this.mapFieldNames(element, stripNulls));
		}
		return outArray;
	}


	this.serverType = 'mssql'; //could be 'file' 'mysql' or 'postgresql' or ...? //maps to something in the dictionary.definition

	//INITIALIZATION ====================================

	this.converterName = '';


	return this;
};

//END OF moduleFunction() ============================================================

util.inherits(moduleFunction, events.EventEmitter);
module.exports = moduleFunction;
